pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageReplicaGroupSelector.java

```diff
@@ -99,58 +100,139 @@ Pair<Map<String, String>, Map<String, String>> select(List<String> segments, int
     } else {
       replicaGroupSelected = requestId % instancePartitions.getNumReplicaGroups();
     }
-    for (int iteration = 0; iteration < instancePartitions.getNumReplicaGroups(); iteration++) {
-      int replicaGroup = (replicaGroupSelected + iteration) % instancePartitions.getNumReplicaGroups();
-      try {
-        return tryAssigning(segments, segmentStates, instancePartitions, replicaGroup);
-      } catch (Exception e) {
-        LOGGER.warn("Unable to select replica-group {} for table: {}", replicaGroup, _tableNameWithType, e);
+
+    return tryAssigning(Sets.newHashSet(segments), segmentStates, instancePartitions, replicaGroupSelected);
+  }
+
+  /**
+   * Returns a map from the segmentName to the corresponding server. It tries to select all servers from the
+   * preferredReplicaGroup, but if it fails, it will try to select the relevant server from other instance partitions.
+   */
+  private Pair<Map<String, String>, Map<String, String>> tryAssigning(
+    Set<String> segments,
+    SegmentStates segmentStates,
+    InstancePartitions instancePartitions,
+    int preferredReplicaId) {
+
+    Map<String, Integer> instanceToPartitionMap = instancePartitions.getInstanceToPartitionIdMap();
+    Map<String, Set<String>> instanceToSegmentsMap = new HashMap<>();
+
+    // instanceToSegmentsMap stores the mapping from instance to the active segments it can serve.
+    for (String segment : segments) {
+      List<SegmentInstanceCandidate> candidates = segmentStates.getCandidates(segment);
+      Preconditions.checkState(candidates != null, "Failed to find servers for segment: %s", segment);
+      for (SegmentInstanceCandidate candidate : candidates) {
```

Comment:
need to handle scenario when `candidates` list is empty. otherwise the segment will be silently ignored